### ISSUE TITLE
400 Bad Request error suppressed when invalid API key is provided

### ISSUE THREAD

**Issue by zdb** [thibaudgg/video_info#86]
VideoInfo::Providers::Youtube.available?  suppresses errors via rescue false.
```
VideoInfo.provider_api_keys = { youtube: 'INVALID_KEY'}
video = VideoInfo.new('http://www.youtube.com/watch?v=mZqGqE0D0n4')
puts video.available? #false
#(400 Bad Request error suppressed)
```

```
VideoInfo.provider_api_keys = {}
video = VideoInfo.new('http://www.youtube.com/watch?v=mZqGqE0D0n4')
puts video.available? #false
#(400 Bad Request error suppressed)
```

```
VideoInfo.provider_api_keys = { youtube: 'VALID_KEY'}
video = VideoInfo.new('http://www.youtube.com/watch?v=mZqGqE0D0n4')
puts video.available? #true
```

**Comment by thibaudgg** [comment]
Master branch?

**Comment by thibaudgg** [comment]
Still an issue? Could you please try again with master branch?

**Comment by vheuken** [comment]
@thibaudgg  I don't think I fixed it. I'll look at it when I handle #62

**Comment by thibaudgg** [comment]
:+1:

**Comment by 0mega** [comment]
For me this only returns false when the API key is invalid. When API key is empty, `#available?`returns true (on master). So probably there is no issue anymore.

**Comment by thibaudgg** [comment]
@zdb do you confirm?

**Comment by vheuken** [comment]
I also get the same behavior that @0mega described.

**Comment by vheuken** [comment]
The behavior @0mega and I experienced is because now with no API key, it just uses the scraper.

The real issue here is that we aren't actually telling the end user that the API key is invalid.

I modified the title from "VideoInfo::Providers::Youtube.available? returns false when API Key not supplied or invalid" to "400 Bad Request error suppressed when invalid API key provided"

**Comment by 0mega** [comment]
I started looking  at this. So the expected behavior seems to be to raise an exception (ex: VideoInfo::ApiKeyError) when user token is supplied and the response from api is 400 (so we assume the token is invalid).

This will affect all methods, not just the `#available?`, so any method which calls `#data` method would raise the VideoInfo::ApiKeyError in this case.
The problem is that I'm sure this will break some clients. Although these clients shouldn't have had useful behavior with an invalid token, the application did not raise error, and now it will.

What do you guys think?

**Comment by vheuken** [comment]
Hm.... I mean, the behavior sounds perfectly reasonable, but I don't like the idea of breaking other people's code on what should be a minor release. 

>Although these clients shouldn't have had useful behavior with an invalid token

Maybe we can have them fall back to the scraper if the API key is invalid? The thing is I'd want to make sure they *know* that they fell back to the scraper. Not sure what the best way to accomplish this is. Any ideas?

**Comment by 0mega** [comment]
@vheuken having a fall back to scraper seems nice.
And we can allow people to configure a logger for video_info in initializer which will log the error message when invalid API key is provided. It would also warn them that scraper is being used instead.

**Comment by vheuken** [comment]
Yeah, that sounds like the best solution. Go for it~

**Comment by 0mega** [comment]
Hi guys, I decided it would be better to simply log a warning when a user provided invalid API key. The other solution was getting a bit messy. I think this is OK, when people provide API key, they want to use the API and not a scraper.

**Comment by vheuken** [comment]
Fixed in PR #110
